File: src/code/alphabeta.java
import java.util.*;

/**
 * Alpha-Beta Pruning Algorithm
 * Game Tree Search Optimization
 * 
 * This implementation demonstrates Alpha-Beta pruning on a fixed game tree.
 */

class GameTreeNode {
    int id;
    String type;  // "MAX", "MIN", or "TERMINAL"
    int value;
    List<GameTreeNode> children;
    int alpha;
    int beta;
    boolean isPruned;
    
    public GameTreeNode(int id, String type, int value) {
        this.id = id;
        this.type = type;
        this.value = value;
        this.children = new ArrayList<>();
        this.alpha = Integer.MIN_VALUE;
        this.beta = Integer.MAX_VALUE;
        this.isPruned = false;
    }
    
    public GameTreeNode(int id, String type) {
        this(id, type, 0);
    }
    
    public void addChild(GameTreeNode child) {
        children.add(child);
    }
}

public class AlphaBetaPruning {
    
    /**
     * Create the predefined game tree
     * Tree structure: MAX -> MIN (2) -> MAX (4) -> MIN (8) -> TERMINAL (16)
     */
    public static GameTreeNode createGameTree() {
        // Root - MAX node
        GameTreeNode root = new GameTreeNode(0, "MAX");
        
        // Level 1 - MIN nodes (2 nodes)
        GameTreeNode min1 = new GameTreeNode(1, "MIN");
        GameTreeNode min2 = new GameTreeNode(2, "MIN");
        root.addChild(min1);
        root.addChild(min2);
        
        // Level 2 - MAX nodes (4 nodes - 2 per MIN)
        GameTreeNode max1 = new GameTreeNode(3, "MAX");
        GameTreeNode max2 = new GameTreeNode(4, "MAX");
        GameTreeNode max3 = new GameTreeNode(5, "MAX");
        GameTreeNode max4 = new GameTreeNode(6, "MAX");
        
        min1.addChild(max1);
        min1.addChild(max2);
        min2.addChild(max3);
        min2.addChild(max4);
        
        // Level 3 - MIN nodes (8 nodes - 2 per MAX)
        List<GameTreeNode> minNodesL3 = new ArrayList<>();
        int nodeId = 7;
        List<GameTreeNode> maxNodes = Arrays.asList(max1, max2, max3, max4);
        
        for (GameTreeNode maxNode : maxNodes) {
            for (int i = 0; i < 2; i++) {
                GameTreeNode minNode = new GameTreeNode(nodeId++, "MIN");
                maxNode.addChild(minNode);
                minNodesL3.add(minNode);
            }
        }
        
        // Level 4 - Terminal nodes (16 nodes - 2 per MIN)
        int[] terminalValues = {3, 4, 2, 1, 7, 8, 9, 10, 2, 11, 1, 12, 14, 9, 13, 16};
        int terminalId = 15;
        int terminalIndex = 0;
        
        for (GameTreeNode minNode : minNodesL3) {
            for (int i = 0; i < 2; i++) {
                int value = terminalValues[terminalIndex++];
                GameTreeNode terminal = new GameTreeNode(terminalId++, "TERMINAL", value);
                minNode.addChild(terminal);
            }
        }
        
        return root;
    }
    
    /**
     * Alpha-Beta Pruning Algorithm
     */
    public static int alphaBetaPruning(GameTreeNode node, int alpha, int beta, 
                                      boolean isMaximizing, int depth, List<String> log) {
        String indent = "  ".repeat(depth);
        
        // Terminal node - return its value
        if (node.type.equals("TERMINAL")) {
            if (log != null) {
                log.add(indent + "Node " + node.id + " (TERMINAL): value = " + node.value);
            }
            return node.value;
        }
        
        if (isMaximizing) {
            // MAX node
            if (log != null) {
                log.add(indent + "Node " + node.id + " (MAX): α=" + alpha + ", β=" + beta);
            }
            
            int value = Integer.MIN_VALUE;
            
            for (int i = 0; i < node.children.size(); i++) {
                if (beta <= alpha) {
                    // Prune remaining children
                    for (int j = i; j < node.children.size(); j++) {
                        node.children.get(j).isPruned = true;
                        if (log != null) {
                            log.add(indent + "  ✂️ PRUNED Node " + node.children.get(j).id + 
                                   ": α(" + alpha + ") ≥ β(" + beta + ")");
                        }
                    }
                    break;
                }
                
                int childValue = alphaBetaPruning(node.children.get(i), alpha, beta, 
                                                 false, depth + 1, log);
                value = Math.max(value, childValue);
                alpha = Math.max(alpha, value);
                
                if (log != null) {
                    log.add(indent + "  Updated: value=" + value + ", α=" + alpha);
                }
            }
            
            node.value = value;
            return value;
        } else {
            // MIN node
            if (log != null) {
                log.add(indent + "Node " + node.id + " (MIN): α=" + alpha + ", β=" + beta);
            }
            
            int value = Integer.MAX_VALUE;
            
            for (int i = 0; i < node.children.size(); i++) {
                if (beta <= alpha) {
                    // Prune remaining children
                    for (int j = i; j < node.children.size(); j++) {
                        node.children.get(j).isPruned = true;
                        if (log != null) {
                            log.add(indent + "  ✂️ PRUNED Node " + node.children.get(j).id + 
                                   ": β(" + beta + ") ≤ α(" + alpha + ")");
                        }
                    }
                    break;
                }
                
                int childValue = alphaBetaPruning(node.children.get(i), alpha, beta, 
                                                 true, depth + 1, log);
                value = Math.min(value, childValue);
                beta = Math.min(beta, value);
                
                if (log != null) {
                    log.add(indent + "  Updated: value=" + value + ", β=" + beta);
                }
            }
            
            node.value = value;
            return value;
        }
    }
    
    /**
     * Collect all pruned nodes
     */
    public static void collectPrunedNodes(GameTreeNode node, List<Integer> prunedList) {
        if (node.isPruned) {
            prunedList.add(node.id);
        }
        for (GameTreeNode child : node.children) {
            collectPrunedNodes(child, prunedList);
        }
    }
    
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("ALPHA-BETA PRUNING VISUALIZER");
        System.out.println("=".repeat(60));
        System.out.println();
        
        // Create the game tree
        GameTreeNode tree = createGameTree();
        System.out.println("Game tree created with 31 total nodes");
        System.out.println("Tree structure: MAX -> MIN (2) -> MAX (4) -> MIN (8) -> TERMINAL (16)");
        System.out.println("Terminal values: 3, 4, 2, 1, 7, 8, 9, 10, 2, 11, 1, 12, 14, 9, 13, 16");
        System.out.println();
        
        // Run Alpha-Beta Pruning
        System.out.println("Running Alpha-Beta Pruning...");
        System.out.println("-".repeat(60));
        
        List<String> log = new ArrayList<>();
        int optimalValue = alphaBetaPruning(tree, 
                                           Integer.MIN_VALUE, 
                                           Integer.MAX_VALUE, 
                                           true, 
                                           0, 
                                           log);
        System.out.println();
        
        // Print execution log
        System.out.println("EXECUTION LOG:");
        System.out.println("-".repeat(60));
        for (String entry : log) {
            System.out.println(entry);
        }
        System.out.println();
        
        // Collect pruned nodes
        List<Integer> prunedNodes = new ArrayList<>();
        collectPrunedNodes(tree, prunedNodes);
        Collections.sort(prunedNodes);
        
        // Print results
        System.out.println("=".repeat(60));
        System.out.println("RESULTS:");
        System.out.println("=".repeat(60));
        System.out.println("✓ Optimal Value (Root): " + optimalValue);
        System.out.println("✓ Pruned Nodes: " + prunedNodes.size());
        System.out.print("  Node IDs: ");
        for (int i = 0; i < prunedNodes.size(); i++) {
            System.out.print(prunedNodes.get(i));
            if (i < prunedNodes.size() - 1) System.out.print(", ");
        }
        System.out.println();
        System.out.printf("✓ Efficiency: %d/31 nodes = %.1f%% reduction%n", 
                         prunedNodes.size(), (prunedNodes.size() / 31.0 * 100));
        System.out.println("=".repeat(60));
    }
}

File: src/code/alphabeta.py
"""
Alpha-Beta Pruning Algorithm
Game Tree Search Optimization

This implementation demonstrates Alpha-Beta pruning on a fixed game tree.
The tree alternates between MAX and MIN levels, with terminal nodes 
containing predefined utility values.
"""

class GameTreeNode:
    """Represents a node in the game tree"""
    def __init__(self, node_id, node_type, value=None):
        self.id = node_id
        self.type = node_type  # 'MAX', 'MIN', or 'TERMINAL'
        self.value = value
        self.children = []
        self.alpha = float('-inf')
        self.beta = float('inf')
        self.is_pruned = False
    
    def add_child(self, child):
        """Add a child node"""
        self.children.append(child)


def create_game_tree():
    """
    Create the predefined game tree from the problem.
    Tree structure: MAX -> MIN (2) -> MAX (4) -> MIN (8) -> TERMINAL (16)
    Leaf values (left to right): 3, 4, 2, 1, 7, 8, 9, 10, 2, 11, 1, 12, 14, 9, 13, 16
    """
    # Root - MAX node
    root = GameTreeNode(0, 'MAX')
    
    # Level 1 - MIN nodes (2 nodes)
    min1 = GameTreeNode(1, 'MIN')
    min2 = GameTreeNode(2, 'MIN')
    root.add_child(min1)
    root.add_child(min2)
    
    # Level 2 - MAX nodes (4 nodes - 2 per MIN)
    max1 = GameTreeNode(3, 'MAX')
    max2 = GameTreeNode(4, 'MAX')
    max3 = GameTreeNode(5, 'MAX')
    max4 = GameTreeNode(6, 'MAX')
    
    min1.add_child(max1)
    min1.add_child(max2)
    min2.add_child(max3)
    min2.add_child(max4)
    
    # Level 3 - MIN nodes (8 nodes - 2 per MAX)
    min_nodes_l3 = []
    node_id = 7
    max_nodes = [max1, max2, max3, max4]
    
    for max_node in max_nodes:
        for i in range(2):
            min_node = GameTreeNode(node_id, 'MIN')
            max_node.add_child(min_node)
            min_nodes_l3.append(min_node)
            node_id += 1
    
    # Level 4 - Terminal nodes (16 nodes - 2 per MIN)
    terminal_values = [3, 4, 2, 1, 7, 8, 9, 10, 2, 11, 1, 12, 14, 9, 13, 16]
    terminal_id = 15
    terminal_index = 0
    
    for min_node in min_nodes_l3:
        for i in range(2):
            terminal = GameTreeNode(terminal_id, 'TERMINAL', terminal_values[terminal_index])
            min_node.add_child(terminal)
            terminal_id += 1
            terminal_index += 1
    
    return root


def alpha_beta_pruning(node, alpha, beta, is_maximizing, depth=0, log=None):
    """
    Alpha-Beta Pruning Algorithm
    
    Args:
        node: Current game tree node
        alpha: Best value for MAX player (initially -∞)
        beta: Best value for MIN player (initially +∞)
        is_maximizing: True if current player is MAX
        depth: Current depth in tree (for logging)
        log: List to store execution log
    
    Returns:
        Optimal value for the current node
    """
    if log is None:
        log = []
    
    indent = "  " * depth
    
    # Terminal node - return its value
    if node.type == 'TERMINAL':
        log.append(f"{indent}Node {node.id} (TERMINAL): value = {node.value}")
        return node.value
    
    if is_maximizing:
        # MAX node
        log.append(f"{indent}Node {node.id} (MAX): α={alpha}, β={beta}")
        value = float('-inf')
        
        for i, child in enumerate(node.children):
            if beta <= alpha:
                # Prune remaining children
                for j in range(i, len(node.children)):
                    node.children[j].is_pruned = True
                    log.append(f"{indent}  ✂️ PRUNED Node {node.children[j].id}: α({alpha}) ≥ β({beta})")
                break
            
            child_value = alpha_beta_pruning(child, alpha, beta, False, depth + 1, log)
            value = max(value, child_value)
            alpha = max(alpha, value)
            
            log.append(f"{indent}  Updated: value={value}, α={alpha}")
        
        node.value = value
        return value
    
    else:
        # MIN node
        log.append(f"{indent}Node {node.id} (MIN): α={alpha}, β={beta}")
        value = float('inf')
        
        for i, child in enumerate(node.children):
            if beta <= alpha:
                # Prune remaining children
                for j in range(i, len(node.children)):
                    node.children[j].is_pruned = True
                    log.append(f"{indent}  ✂️ PRUNED Node {node.children[j].id}: β({beta}) ≤ α({alpha})")
                break
            
            child_value = alpha_beta_pruning(child, alpha, beta, True, depth + 1, log)
            value = min(value, child_value)
            beta = min(beta, value)
            
            log.append(f"{indent}  Updated: value={value}, β={beta}")
        
        node.value = value
        return value


def collect_pruned_nodes(node, pruned_list):
    """Collect all pruned nodes in the tree"""
    if node.is_pruned:
        pruned_list.append(node.id)
    for child in node.children:
        collect_pruned_nodes(child, pruned_list)


def main():
    """Main execution"""
    print("=" * 60)
    print("ALPHA-BETA PRUNING VISUALIZER")
    print("=" * 60)
    print()
    
    # Create the game tree
    tree = create_game_tree()
    print("Game tree created with 31 total nodes")
    print("Tree structure: MAX -> MIN (2) -> MAX (4) -> MIN (8) -> TERMINAL (16)")
    print("Terminal values: 3, 4, 2, 1, 7, 8, 9, 10, 2, 11, 1, 12, 14, 9, 13, 16")
    print()
    
    # Run Alpha-Beta Pruning
    print("Running Alpha-Beta Pruning...")
    print("-" * 60)
    log = []
    optimal_value = alpha_beta_pruning(
        tree, 
        float('-inf'), 
        float('inf'), 
        True,  # Root is MAX node
        log=log
    )
    print()
    
    # Print execution log
    print("EXECUTION LOG:")
    print("-" * 60)
    for entry in log:
        print(entry)
    print()
    
    # Collect pruned nodes
    pruned_nodes = []
    collect_pruned_nodes(tree, pruned_nodes)
    
    # Print results
    print("=" * 60)
    print("RESULTS:")
    print("=" * 60)
    print(f"✓ Optimal Value (Root): {optimal_value}")
    print(f"✓ Pruned Nodes: {len(pruned_nodes)}")
    print(f"  Node IDs: {sorted(pruned_nodes)}")
    print(f"✓ Efficiency: {len(pruned_nodes)}/31 nodes = {(len(pruned_nodes)/31)*100:.1f}% reduction")
    print("=" * 60)


if __name__ == "__main__":
    main()
